nn/nntree.py

```diff
@@ -0,0 +1,499 @@
+def flat_is_left_nonterminal(string):
+    return flat_is_nonterminal(string) and not string.startswith("/")
+
+
+def flat_is_right_nonterminal(string):
+    return flat_is_nonterminal(string) and string.startswith("/")
+
+
+def flat_matching_nonterminal(string):
+    if string.startswith("/"):
+        return string[1:]
+    return "/" + string
+
+
+class Node:
+    """ Generic tree implementation """
+
+    def __init__(self, name, is_terminal=False, data=None):
+        self.name = name
+        self.data = data
+        self.children = []
+        self.is_terminal = is_terminal
+
+    def add_child(self, node):
+        self.children.append(node)
+        return self
+
+    def has_children(self):
+        return bool(self.children)
+
+    def to_dict(self):
+        if not self.children:
+            json_node = _json_terminal_node(self.name, self.data)
+        else:
+            json_node = _json_nonterminal_node(self.name)
+            json_node[KEY.children] = [c.to_dict() for c in self.children]
+        return json_node
+
+    def to_simple_tree(self):
+        return matcher.SimpleTree([[self.to_dict()]])
+
+    def to_postfix(self):
+        """ Export tree to postfix ordering
+            with node-labels as keys """
+        result = []
+        for child in self.children:
+            result.extend(child.to_postfix())
+        result.append(self.name)
+        return result
+
+    def __str__(self):
+        text = (" " + self.data) if self.data is not None else ""
+        return "<" + self.name + text + ">"
+
+    def __repr__(self):
+        if self.is_terminal:
+            return self.__str__()
+        result = [self.__str__()]
+        for child in self.children:
+            result.extend(child.__repr__())
+        result.append(Node("/" + self.name).__str__())
+        return "".join(result)
+
+    def _pprint(self, _prefix="  ", depth=0):
+        """ Inner function for pretty printing tree """
+        print("{}{}".format(_prefix * depth, self.__str__()))
+        for c in self.children:
+            c._pprint(_prefix=_prefix, depth=depth + 1)
+
+    def pprint(self, indent=4):
+        """ Pretty print tree """
+        self._pprint(_prefix=" " * indent)
+
+    def width(self):
+        """ Returns number of leaf nodes,
+            assumes a correctly formed tree """
+        if self.children:
+            return sum(c.width() for c in self.children)
+        return 1
+
+    def height(self):
+        """ Returns height of tree,
+            assumes a correctly formed tree """
+        if not self.children:
+            return 1
+        return 1 + max(c.height() for c in self.children)
+
+    @staticmethod
+    def contains(forest, name):
+        if isinstance(forest, Node):
+            forest = [forest]
+        for tree in forest:
+            if tree.name == name:
+                return True
+            if any(Node.contains(c, name) for c in tree.children):
+                return True
+        return False
+
```

Comment:
Extra empty line here

Reply:
There is a new class defined below.

Reply:
I think I saw three consecutive empty lines, which should have been two, but I might be wrong.